File: 99. Recover Binary Search Tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def recoverTree(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        nodes = list()
        self.inorder(root, None, nodes)
        if nodes:
            nodes[0].val, nodes[1].val = nodes[1].val, nodes[0].val
            
    def inorder(self, root, prev, nodes):
        if not root:
            return prev
        
        prev = self.inorder(root.left, prev, nodes)
        if prev and prev.val > root.val:
            if not nodes:
                nodes.append(prev)
                nodes.append(root)
            else:
                nodes.pop()
                nodes.append(root)
                
        return self.inorder(root.right, root, nodes)

File: 99. Recover Binary Search Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void recoverTree(TreeNode root) {
        TreeNode[] buffer = new TreeNode[2];
        buffer[0] = null;
        buffer[1] = null;
        
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode prev = null;
        TreeNode node = root;
        
        // inorder traversal to find the non increasing order
        while (node != null || !stack.isEmpty()) {
            if (node != null) {
                stack.add(node);
                node = node.left;
            } else {
                node = stack.pop();
                if (node == null) {
                    continue;
                }
                
                if (prev != null && prev.val > node.val) {
                    if (buffer[0] == null) {
                        buffer[0] = prev;
                    }
                    buffer[1] = node;
                }
                
                prev = node;
                node = node.right;
            }
        }
        
        int tmp;
        tmp = buffer[0].val;
        buffer[0].val = buffer[1].val;
        buffer[1].val = tmp;
    }
}
